File: client/src/test/interruptWindow.test.ts
import { describe, it, expect, beforeEach } from 'vitest'
import { createServerMessageHandler } from '../hooks/serverMessages'
import { gameStore } from '../hooks/gameStore'
import type { CardDTO } from '../types/protocol'

/**
 * Whether the pile may still be slammed is the server's word, carried on every
 * message that can open or shut the window. The client used to keep no copy of
 * it and offered the twin for as long as the card was on top, so a slam after
 * somebody had drawn came back "somebody was faster" on a table where nobody
 * had been.
 */
const red5: CardDTO = { color: 'red', kind: 'number', value: 5 }
const roster = [
  { index: 0, nickname: 'Me', hand_size: 3, connected: true },
  { index: 1, nickname: 'Them', hand_size: 4, connected: true },
]

beforeEach(() => {
  gameStore.setState({ myIndex: 0, players: roster, interruptOpen: true, catchWindows: [] })
})

describe('the interrupt window in the store', () => {
  const handle = createServerMessageHandler({ clear: () => {}, arm: () => {} })

  it('shuts when the seat at turn draws, and not when a hand merely grows', () => {
    handle({ type: 'card_drawn', player_index: 1, turn: 1, drawn_count: 1, interrupt_open: false })
    expect(gameStore.getState().interruptOpen).toBe(false)
    handle({ type: 'card_played', player_index: 1, card: red5, turn: 0, players: roster })
    expect(gameStore.getState().interruptOpen).toBe(true)
    // A penalty draw says nothing about the window, and the server sends it
    // as open: the value on the message is what counts, never the message type.
    handle({ type: 'card_drawn', player_index: 1, turn: 0, drawn_count: 2, interrupt_open: true })
    expect(gameStore.getState().interruptOpen).toBe(true)
  })

  it('shuts on a pass and reopens on the next card', () => {
    handle({ type: 'turn_changed', turn: 1, interrupt_open: false })
    expect(gameStore.getState().interruptOpen).toBe(false)
    handle({ type: 'card_played', player_index: 1, card: red5, turn: 0, players: roster, interrupt_open: true })
    expect(gameStore.getState().interruptOpen).toBe(true)
  })

  it('takes the snapshot\'s answer, and a snapshot that says nothing is shut', () => {
    handle({ type: 'turn_changed', turn: 1 })
    expect(gameStore.getState().interruptOpen).toBe(false)
    handle({
      type: 'game_state',
      state: {
        your_index: 0,
        hand: [],
        players: roster,
        discard: red5,
        active_color: 'red',
        turn: 0,
        direction: 1,
        round_number: 1,
        match_format: 'BO1',
        max_players: 2,
        interrupt_open: true,
      },
    })
    expect(gameStore.getState().interruptOpen).toBe(true)
    handle({
      type: 'game_state',
      state: {
        your_index: 0,
        hand: [],
        players: roster,
        discard: red5,
        active_color: 'red',
        turn: 0,
        direction: 1,
        round_number: 1,
        match_format: 'BO1',
        max_players: 2,
      },
    })
    expect(gameStore.getState().interruptOpen).toBe(false)
  })
})

File: server/hub/interrupt_window_test.go
package hub_test

import (
	"testing"

	"loco/server/protocol"
)

// Whether the pile may still be slammed is the server's word, and it rides
// every message that can open or shut the window. The client kept no copy of it
// and offered the twin for as long as the card was on top, so a slam after the
// seat at turn had drawn was answered "interrupt window closed" and rendered as
// "somebody was faster" on a table where nobody had been.
func TestInterruptOpen_RidesTheWire(t *testing.T) {
	t.Setenv("LOCO_E2E", "1")
	_, srv := newTestHub(t)
	conn1, conn2, _ := setupTwoPlayerGame(t, srv)

	// Alice at turn holding two red numbers on a red top; nothing pending.
	red7 := protocol.CardDTO{Color: "red", Kind: "number", Value: 7}
	red3 := protocol.CardDTO{Color: "red", Kind: "number", Value: 3}
	top := protocol.CardDTO{Color: "red", Kind: "number", Value: 9}
	zero, turn := 0, 0
	sendMsg(t, conn1, protocol.ClientMsg{
		Type: protocol.CMsgDebugSetState,
		Debug: &protocol.DebugStateDTO{
			Hand:        []protocol.CardDTO{red7, red3},
			Discard:     &top,
			PendingDraw: &zero,
			CurrentTurn: &turn,
		},
	})
	snap := readMsgOfType(t, conn1, protocol.SMsgGameState)
	readMsgOfType(t, conn2, protocol.SMsgGameState)
	if snap.State == nil || !snap.State.InterruptOpen {
		t.Fatalf("the snapshot of a fresh deal says the window is shut")
	}

	// A play opens it, and says so.
	sendMsg(t, conn1, protocol.ClientMsg{Type: protocol.CMsgPlayCard, Card: &red7})
	played := readMsgOfType(t, conn2, protocol.SMsgCardPlayed)
	if played.InterruptOpen == nil || !*played.InterruptOpen {
		t.Fatalf("card_played interrupt_open = %v, want true", played.InterruptOpen)
	}

	// Bob, at turn now, draws: the window shuts and every seat is told.
	sendMsg(t, conn2, protocol.ClientMsg{Type: protocol.CMsgDrawCard})
	drawn := readMsgOfType(t, conn1, protocol.SMsgCardDrawn)
	if drawn.InterruptOpen == nil || *drawn.InterruptOpen {
		t.Fatalf("card_drawn interrupt_open = %v, want false after the seat at turn drew", drawn.InterruptOpen)
	}

	// And the pass that follows carries the same answer.
	sendMsg(t, conn2, protocol.ClientMsg{Type: protocol.CMsgPassTurn})
	changed := readMsgOfType(t, conn1, protocol.SMsgTurnChanged)
	if changed.InterruptOpen == nil || *changed.InterruptOpen {
		t.Fatalf("turn_changed interrupt_open = %v, want false", changed.InterruptOpen)
	}
}
